CMP extension to OpenSSL, incremental PR chunk 2: CRMF code

DDvO opened the pull request:
Certificate Management Protocol (CMP, RFC 4210) extension to OpenSSL.
Also includes CRMF (RFC 4211) and HTTP transfer (RFC 6712).

CMP and CRMF functionality is added to `libcrypto`, and the `cmp` app to the `openssl` CLI.
Adds extensive man pages and tests. Integration into build scripts.

Incremental pull request as suggested by @mattcaswell: #6811 (comment).

Initial chunk, reviewed and approved in #7328, consists of the CRMF API
as declared in the header file `include/openssl/crmf.h`
and its documentation in `doc/man3/OSSL_CRMF_*.pod`

This second chunk adds the CRMF implementation (`crypto/crmf/`)
and integrates it in `INSTALL`, `Configure`, `crypto/build.info`, `crypto/err/err.c`, `crypto/err/err_all.c`, `crypto/err/openssl.ec`, `crypto/err/openssl.txt`, `include/openssl/err.h`, `include/openssl/crmferr.h`, and `util/libcrypto.num`.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [ ] tests are added or updated

bernd-edlinger commented on crypto/crmf/crmf_pbm.c:
```diff
@@ -0,0 +1,222 @@
+        error = CRMF_R_UNSUPPORTED_ALGORITHM;
+        goto err;
+    }
+    HMAC(m, basekey, bklen, msg, msglen, *mac, maclen);
```
return code from HMAC is ignored.

DDvO replied:
Fixed

bernd-edlinger commented on crypto/crmf/crmf_pbm.c:
```diff
@@ -0,0 +1,222 @@
+
+    /* cleanup */
+    OPENSSL_cleanse(basekey, bklen);
+    EVP_MD_CTX_destroy(ctx);
```
ctx not cleaned up in error handler

DDvO replied:
Fixed, also cleansing of basekey containing some secret data

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,769 @@
+        goto err;
+    }
+
+    ASN1_INTEGER_set(pi->action, action);
```
return code ignored.

DDvO replied:
Fixed

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,769 @@
+
+    res = ASN1_INTEGER_set(crm->certReq->certReqId, rid);
+    if (res == 0)
+        CRMFerr(CRMF_F_OSSL_CRMF_MSG_SET_CERTREQID, CRMF_R_NULL_ARGUMENT);
```
this is more likely an out of memory?

DDvO replied:
Yes, fixed

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,769 @@
+    if ((p = outbuf = OPENSSL_malloc(ecert->encValue->length +
+                                     EVP_CIPHER_block_size(cipher))) == NULL)
+        goto oom;
+    evp_ctx = EVP_CIPHER_CTX_new();
```
check for null missing here.

DDvO replied:
Fixed

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,769 @@
+        goto err;                                                         \
+    return 1;                                                             \
+ err:                                                                     \
+    OSSL_CRMF_ATTRIBUTETYPEANDVALUE_free(atav);                 \
```
sloppy indentation

DDvO replied:
improved

bernd-edlinger commented on crypto/build.info:
```diff
@@ -5,6 +5,7 @@ SUBDIRS=objects buffer bio stack lhash rand evp asn1 pem x509 x509v3 conf \
         md2 md4 md5 sha mdc2 gmac hmac ripemd whrlpool poly1305 blake2 \
         siphash sm3 des aes rc2 rc4 rc5 idea aria bf cast camellia \
         seed sm4 chacha modes bn ec rsa dsa dh sm2 dso engine \
+        crmf \
```
merge this with the following one.

DDvO replied:
Done. I had put 'crmf' in an extra line such to stress that this is an addition and to avoid touching an existing line.

bernd-edlinger commented on crypto/crmf/crmf_int.h:
```diff
@@ -0,0 +1,401 @@
+# include <openssl/x509v3.h>
+
+# ifdef  __cplusplus
+extern "C" {
```
internal headers don't need this.

DDvO replied:
Good to know; removed.

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,769 @@
+    X509_ALGOR_set0(ps->algorithmIdentifier, OBJ_nid2obj(alg_nid),
+                    V_ASN1_NULL, NULL);
+
+    ctx = EVP_MD_CTX_create();
```
check for ctx==null here.

DDvO replied:
Done

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,769 @@
+    OSSL_CRMF_POPOSIGNINGKEY_free(ps);
+    OPENSSL_free(crder);
+    if (ctx)
+        EVP_MD_CTX_destroy(ctx);
```
EVP_MD_CTX_create() and EVP_MD_CTX_destroy() were renamed to
EVP_MD_CTX_new() and EVP_MD_CTX_free() in OpenSSL 1.1.0.

DDvO replied:
Updated

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,769 @@
+                CRMF_R_UNSUPPORTED_ALG_FOR_POPSIGNINGKEY);
+        goto err;
+    }
+    X509_ALGOR_set0(ps->algorithmIdentifier, OBJ_nid2obj(alg_nid),
```
return code needs checking.

DDvO replied:
Fixed

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,769 @@
+    ps->signature->flags &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07);
+    ps->signature->flags |= ASN1_STRING_FLAG_BITS_LEFT;
+
+    l = i2d_OSSL_CRMF_CERTREQUEST(cr, &crder);
```
I don't like this identifier, please don't use lower-case L, because I can't see the difference to 1.

DDvO replied:
Yes; changed to 'len'.

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,769 @@
+            != OSSL_CRMF_POPOPRIVKEY_SUBSEQUENTMESSAGE)
+            goto unsupported;
+        if (ASN1_INTEGER_get
+            (req->popo->value.keyEncipherment->value.subsequentMessage) !=
```
!= goes to the next line

DDvO replied:
Done

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,769 @@
+            OSSL_CRMF_SUBSEQUENTMESSAGE_ENCRCERT)
+            goto unsupported;
+#if 0 /* TODO enable code when implemented in CMP_certrep_new() */
+        srv_ctx->encryptcert = 1;
```
we usually don't have commented code. especially when it would not even compile (no srv_ctx anywhere)

DDvO replied:
Yes, part of the TODO should have been place to some other source file (`crypto/cmp/cmp_srv.c`).
Done this and change the rest including the `#if` to a simple TODO comment.

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,769 @@
+        CRMFerr(CRMF_F_OSSL_CRMF_CERTTEMPLATE_FILL, CRMF_R_NULL_ARGUMENT);
+        return 0;
+    }
+    if ((pubkey != NULL &&
```
&& goes to the next line, usually with 4 spaces extra indentation per our coding style guide

bernd-edlinger replied:
in this case, probably better to split the expression in several if(...) goto oom?

mattcaswell replied:
The coding style needs updating in various ways. This rule should be added (because we do apply that style).

DDvO replied:
This was also done in November:
```
commit 2f301f13961ad855ec36596092c798b763990684
Author: David von Oheimb <David.von.Oheimb@siemens.com>
Date:   Mon Nov 26 09:58:42 2018 +0100

    fix various minor conding style issues (regarding layout etc.)
```

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,769 @@
+                    CRMF_R_ERROR_DECRYPTING_SYMMETRIC_KEY);
+            goto end;
+        }
+    } else
```
this is against our coding style. 
```
   } else {
       goto oom;
   }
```

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,769 @@
+    int n, outlen = 0;
+    EVP_PKEY_CTX *pkctx = NULL; /* private key context */
+
+    if (ecert == NULL || ecert->symmAlg == NULL || ecert->encSymmKey == NULL ||
```
operators at end of line go to the next line

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,769 @@
+    }
+    if ((iv = OPENSSL_malloc(EVP_CIPHER_iv_length(cipher))) == NULL)
+        goto oom;
+    ASN1_TYPE_get_octetstring(ecert->symmAlg->parameter, iv,
```
maybe check the return code?

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,769 @@
+
+    /* first the symmetric key needs to be decrypted */
+    if ((pkctx = EVP_PKEY_CTX_new(pkey, NULL)) != NULL &&
+        EVP_PKEY_decrypt_init(pkctx)) {
```
also here the && goes to the next line, with 4 space indentation:
```
if ((...) != NULL
        && EVP_PKEY_decrypt_init(pkctx)) {
    ASN1_BIT_STRING *encKey = ecert->encSymmKey;
```

FdaSilvaYY commented on crypto/crmf/crmf_int.h:
```diff
@@ -0,0 +1,401 @@
+ * encValue            BIT STRING }
+ *                  -- the encrypted value itself
+ */
+struct OSSL_crmf_encrypetedvalue_st {
```
OSSL_crmf_encryp**e**tedvalue_st

DDvO replied:
Good spotting; fixed (also in `include/openssl/crmf.h`)

FdaSilvaYY commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,769 @@
+}
+
+ /*
+  * id-regCtrl-protocolEncrKey Control (section 6.6) */
```
Someone has hesitate between single or multiple lines comments ? :)

DDvO replied:
I guess this was due to a glitch in automatic re-formatting ;)

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,769 @@
+    if (b == NULL)
+        return 1;
+    (void)X509_PUBKEY_get0_param(NULL, NULL, NULL, &algA, a);
+    (void)X509_PUBKEY_get0_param(NULL, NULL, NULL, &algB, b);
```
Done

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,769 @@
+        return 1;
+    (void)X509_PUBKEY_get0_param(NULL, NULL, NULL, &algA, a);
+    (void)X509_PUBKEY_get0_param(NULL, NULL, NULL, &algB, b);
+    if ((res = X509_ALGOR_cmp(algA, algB)) != 0)
```
NULL checks added

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,769 @@
+    (void)X509_PUBKEY_get0_param(NULL, NULL, NULL, &algB, b);
+    if ((res = X509_ALGOR_cmp(algA, algB)) != 0)
+        return res;
+    return EVP_PKEY_cmp(X509_PUBKEY_get0(a), X509_PUBKEY_get0(b));
```
Done

mattcaswell replied:
Done?? I don't see it here?

DDvO replied:
This was done in 

```
commit 413ed461c4edbd0b0ee5ac338361b962776ec8c4
Author: David von Oheimb <David.von.Oheimb@siemens.com>
Date:   Tue Nov 20 22:04:12 2018 +0100

    fix several missing checks for error conditions in crmf_lib.c
```
(but possibly it was pushed later).

DDvO commented:
@bernd-edlinger and @FdaSilvaYY, thank you for your review comments. 
I've just reacted to all of them and done the respective improvements.

DDvO commented:
Oops, my most recent commit (and push) of various coding-style related fixes somehow did not work last Wednesday; done now.

DDvO commented on crypto/crmf/crmf_pbm.c:
```diff
@@ -0,0 +1,225 @@
+    int error = CRMF_R_CRMFERROR;
+
+    if (mac == NULL || pbmp == NULL || pbmp->mac == NULL ||
+            pbmp->mac->algorithm == NULL || msg == NULL || sec == NULL) {
```
This is fixed meanwhile.

DDvO commented on crypto/crmf/crmf_pbm.c:
```diff
@@ -0,0 +1,225 @@
+
+/*
+ * creates and initializes OSSL_CRMF_PBMPARAMETER (section 4.4)
+ * slen SHOULD be > 8    (16 is common)
```
Done. 
BTW, also regarding documentation of function parameters I did not find any guidance in [the Coding Style document](https://www.openssl.org/policies/codingstyle.html).

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,770 @@
+        pubkey = req->certReq->certTemplate->publicKey;
+        sig = req->popo->value.signature;
+        if (sig->poposkInput != NULL) {
+/* According to RFC 4211:
```
Fixed, moving to the preferred style for long (multi-line) comments :)

DDvO commented on crypto/crmf/crmf_pbm.c:
```diff
@@ -0,0 +1,225 @@
+
+/*
+ * calculates the PBM based on the settings of the given OSSL_CRMF_PBMPARAMETER
+ * @pbmp identifies the algorithms, salt to use
```
Done, also for the remaining parameters

DDvO commented on doc/man3/OSSL_CRMF_MSG_set1_regCtrl_regToken.pod:
```diff
@@ -0,0 +1,105 @@
+OSSL_CRMF_MSG_PKIPublicationInfo_push0_SinglePubInfo() pushes the given B<spi>
+to B<si>. Consumes the B<spi> pointer.
+
+OSSL_CRMF_MSG_set0_SinglePubInfo() sets given B<method> and PubLoction B<nm> to
```
pubLocation is the field name defined in RFC 4211. Nevertheless I've streamlined the documentation at this point.

DDvO commented on doc/man3/OSSL_CRMF_MSG_set1_regCtrl_regToken.pod:
```diff
@@ -0,0 +1,105 @@
+ # define OSSL_CRMF_PUB_ACTION_DONTPUBLISH   0
+ # define OSSL_CRMF_PUB_ACTION_PLEASEPUBLISH 1
+
+OSSL_CRMF_MSG_set1_regCtrl_pkiPublicationInfo() sets the pkiPulicationInfo
```
'pkiPulicationInfo' is the field name defined in RFC 4211, which should be clear from the context:

> OSSL_CRMF_MSG_set1_regCtrl_pkiPublicationInfo() sets the pkiPulicationInfo
> control in the given B<msg> copying the given B<tok> as value. See RFC 4211, section 6.3.

DDvO replied:
So I had misunderstood the original comment. Typo is fixed now.

DDvO commented on doc/man3/OSSL_CRMF_pbmp_new.pod:
```diff
@@ -0,0 +1,78 @@
+malformed input.
+
+OSSL_CRMF_pbmp_new() initializes and returns a new PBMParameter
+structure with new a random salt of given length B<saltlen>, OWF (one-way
```
Fixed

DDvO commented on doc/man3/OSSL_CRMF_pbmp_new.pod:
```diff
@@ -0,0 +1,78 @@
+
+=head1 NOTES
+
+The OWF (one-way function) and for the MAC (message authentication code) may be
```
Reworked the sentence

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,770 @@
+    return tmpl != NULL ? tmpl->issuer : NULL;
+}
+
+/* fill in certificate template.
```
Fixed

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,770 @@
+ * TODO: also support cases 1+2 defined in RFC4211, section 4.1.
+ * returns pointer to created OSSL_CRMF_POPOSIGNINGKEY on success, NULL on error
+ */
+static OSSL_CRMF_POPOSIGNINGKEY *CRMF_poposigkey_new(OSSL_CRMF_CERTREQUEST *cr,
```
Indeed. I moved the allocation of `OSSL_CRMF_POPOSIGNINGKEY *ps` out of `CRMF_poposigkey_new()` and renamed the function to `CRMF_poposigkey_init()`, which is more appropriate.

DDvO commented:
I've meanwhile answered all new comments and pushed respective fixes.

DDvO commented:
@bernd-edlinger and @FdaSilvaYY, are the improvements that I committed fine by you? Or are there more things to do?

HBrock commented:
@mattcaswell is there any further feedback on this CRMF code from your side? 
The goal of these first two pull requests was to get a better understanding of the code quality and the effort for reviewing it on both sides.
What is your suggestion on how to go forward?

mattcaswell commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+        OSSL_CRMF_POPO_free(pp);
+        OSSL_CRMF_POPO_free(crm->popo);
+        crm->popo = NULL;
+        return 1;
```
Perhaps this case could be a special case `if` before we do the memory allocation above. It seems incorrect to allocate and then immediately deallocate.

DDvO replied:
Good point; done.

mattcaswell commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+
+
+int OSSL_CRMF_MSG_create_popo(OSSL_CRMF_MSG *crm, const EVP_PKEY *pkey,
+                              int dgst, int ppmtd)
```
What is `dgst` here? Should this really be an `EVP_MD *`?

DDvO replied:
It's the NID of the desired digest. We found it more convenient to pass around than `EVP_MD *`.

mattcaswell commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+        !X509_PUBKEY_get0_param(NULL, NULL, NULL, &algB, b))
+        return -2;
+    if (algA == NULL || algB == NULL)
+        return -3;
```
I'd don't like the magic number return values. Better would be an enum or macro constants.

mpeylo replied:
The whole static function could be simplified.

DDvO replied:
I've simplified the result values of `X509_PUBKEY_cmp()` to the typical choice of {-1, 0, 1}.

mattcaswell commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+        return 0;
+    }
+
+    switch (req->popo->type) {
```
Can `req->popo` ever by NULL?

mpeylo replied:
probably it could...

DDvO replied:
Indeed it could, since the field is OPTIONAL. I've added a NULL check.

mattcaswell commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+        break;
+    case OSSL_CRMF_POPO_SIGNATURE:
+        pubkey = req->certReq->certTemplate->publicKey;
+        sig = req->popo->value.signature;
```
Could `sig` be NULL?

mpeylo replied:
probably it could...

DDvO replied:
No, because the `signature` field is mandatory.

mattcaswell commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+            != OSSL_CRMF_SUBSEQUENTMESSAGE_ENCRCERT)
+            goto unsupported;
+        /* TODO when implemented in CMP_certrep_new(): return 1 */
+        goto unsupported;
```
This whole case doesn't seem to do anything? Or will this be filled in later by a subsequent PR?

mpeylo replied:
yes, could be removed as there is anyway the fall-through

mattcaswell commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+        goto oom;
+    if (subject != NULL && !X509_NAME_set(&tmpl->subject, (X509_NAME *)subject))
+        goto oom;
+    if (issuer != NULL && !X509_NAME_set(&tmpl->issuer, (X509_NAME *)issuer))
```
Are all the casts here safe? It's not clear to me if they are or not. I started to look but rapidly started going down a rabbit hole. Either we need to fix the underlying functions to declare themselves as taking const params, and explore the rabbit hole properly - or we have to assume that it isn't safe.

DDvO replied:
The idea was that all structures pointed to by parameters (except the first one) of `OSSL_CRMF_CERTTEMPLATE_fill()` should not be touched and this should be reflected by declaring them `const *`. Yet it turns out that `X509_PUBKEY_set()` for whatever reason increments the reference counter of the public key rather than copying the structure. The other parameters can be safely considered constant because `X509_NAME_set()` and `ASN1_INTEGER_dup()` actually do not touch them. Since I found it rather unfortunate that these functions are not (yet) constified I started doing that.

In the end I've constified all *`_dup()` functions as far as possible, as well as most of *`i2d_`*`()` and related functions. This turned out to be quite some amount of work, hopefully to the benefit of many. Among others, I had to introduce a couple of small helper functions and declarations deep inside the ASN.1 implementations.

FdaSilvaYY replied:
I fully support the API constification you made, but it deserves its own PR, IMHO, as it is large patch and not dependent of CRMF support added in this PR.

DDvO replied:
I agree. How best to deal with the problem that this (CRMF) PR will depend on that API constification PR?

mspncp replied:
Create a separate topic branch for the constification, cherry-pick the constification commit, and create a pull request for it.

Then rebase this branch on top of the constification topic branch. Git should notice automatically that the constification commit was already applied. Alternatively, do an interactive rebase (`git rebase -i`) and remove the constification commit from the rebase-todo-list.

You should add a note to the description of this pr that it is based on the other pr, and optionally add the `WIP: ` prefix to the pr title while it's not based onto master.

After the other constification pull request is accepted and merged, you can rebase back onto master.

If you need more technical assistance with the details, just drop me a note.

HTH,
Matthias

DDvO replied:
Thanks @mspncp for this hit!
I'll follow it as soon as this PR has been rebased on the current master.

mattcaswell commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+        ASN1_BIT_STRING *encKey = ecert->encSymmKey;
+        size_t eksize = 0;
+
+        if (EVP_PKEY_decrypt(pkctx, NULL, &eksize, encKey->data, encKey->length)
```
Is any specific RSA padding specified for this?

mpeylo replied:
I will look into that

DDvO replied:
@mpeylo, when are you going to do this?

mpeylo replied:
I spent half an hour searching but couldn't find anything specified related to RSA padding for this, so I came to the conclusion that there with a high likelihood, there is nothing specified.

mattcaswell commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+    if ((iv = OPENSSL_malloc(EVP_CIPHER_iv_length(cipher))) == NULL)
+        goto oom;
+    if (!ASN1_TYPE_get_octetstring(ecert->symmAlg->parameter, iv,
+                                   EVP_CIPHER_iv_length(cipher)))
```
What happens if the iv length is less than expected? Won't it fail in the EVP_DecryptInit call?

mpeylo replied:
I will look into that

DDvO replied:
@mpeylo, when are you going to look into this?

mpeylo replied:
If the IV retrieved from the ASN.1 is too short the data in _iv_ is likely to be wrong. Without having attempted to reproduce this in practice, I doubt that EVP_DecryptInit would fail as the memory size of _iv_ is appropriate, only with some 0-bytes at the end. However, it is likely that the decryption would fail in the end unless the content _iv_ corresponds to what was used on the peer side as IV for encyrypting.
So, in principle, such malformed IV could be explicitly caught as an explicit error with appropriate debug information - whether it is strictly needed I am not sure. However, I'd anyway touch that to put some debugging output before the "goto end" in line 728.

mattcaswell commented on crypto/crmf/crmf_pbm.c:
```diff
@@ -0,0 +1,222 @@
+    int error = CRMF_R_CRMFERROR;
+
+    if ((pbm = OSSL_CRMF_PBMPARAMETER_new()) == NULL) {
+        error = ERR_R_MALLOC_FAILURE;
```
Better is to call CRMFerr here instead. That way in the event of an error we know the actual line number where it occurred. Similarly throughout this function.

DDvO replied:
Done, also in `OSSL_CRMF_pbm_new()`.

mattcaswell commented on crypto/crmf/crmf_pbm.c:
```diff
@@ -0,0 +1,222 @@
+        error = CRMF_R_NULL_ARGUMENT;
+        goto err;
+    }
+    OPENSSL_free(*mac);
```
I'm not sure this is wise. We don't own this memory and didn't allocate it. We have no business freeing it. Things might get confused if it wasn't allocated via OPENSSL_malloc().

DDvO replied:
Good point; I've re-worked that.

mpeylo commented on INSTALL:
```diff
@@ -299,6 +299,9 @@
                    is left enabled (the default), then compression will only
                    work if the zlib or zlib-dynamic options are also chosen.
 
+  no-crmf
```
There are standardized uses cases for CRMF other than CMP. RFC 5272, "Certificate management over CMS" utilizes either PKCS#10 or CRMF. 

If it's the strategy to reduce those options (didn't look like that, but I understand that e.g. run-checker takes quite some time), then it would be easy to reduce and have exclusively no-cmp for both for now.

DDvO replied:
I'll remove the mention of the `no-crmf` option from INSTALL.

Since the implementation is in two directories: `crypto/crmf/` and `crypto/cmp/` I presume that we need to mention both inside the `Configure` script and this will lead to potentially both macros: `OPENSSL_NO_CRMF` and `OPENSSL_NO_CMP`. Since moreover `"cmp"` implies `"crmf"` (which is going to be reflected in `Configure`, I also had to replace all `#ifndef OPENSSL_NO_CRMF` 
by `#if !defined(OPENSSL_NO_CRMF) && !defined(OPENSSL_NO_CMP)`

mpeylo commented on crypto/crmf/crmf_asn.c:
```diff
@@ -0,0 +1,239 @@
+ASN1_CHOICE(OSSL_CRMF_POPO) = {
+    ASN1_IMP(OSSL_CRMF_POPO, value.raVerified, ASN1_NULL, 0),
+    ASN1_IMP(OSSL_CRMF_POPO, value.signature, OSSL_CRMF_POPOSIGNINGKEY, 1),
+    ASN1_EXP(OSSL_CRMF_POPO, value.keyEncipherment, OSSL_CRMF_POPOPRIVKEY, 2),
```
right, needs to be ASN1_IMP

KeyEncipherment POP is not implemented [yet](https://github.com/mpeylo/cmpossl/issues/66), so that hasn't surfaced

mattcaswell replied:
Ah! `OSSL_CRMF_POPOPRIVKEY` is itself a CHOICE type!! So yes `keyAgreement` should also be `ASN1_EXP`.

DDvO replied:
This has already been done in commit 866fac6d, which I pushed just before Christmas.

mpeylo commented on crypto/crmf/crmf_asn.c:
```diff
@@ -0,0 +1,239 @@
+
+
+ASN1_CHOICE(OSSL_CRMF_ENCKEYWITHID_IDENTIFIER) = {
+    ASN1_IMP(OSSL_CRMF_ENCKEYWITHID_IDENTIFIER, value.string, ASN1_UTF8STRING, 0),
```
yes, probably

mpeylo commented on crypto/crmf/crmf_asn.c:
```diff
@@ -0,0 +1,239 @@
+
+ASN1_SEQUENCE(OSSL_CRMF_ENCKEYWITHID) = {
+    ASN1_SIMPLE(OSSL_CRMF_ENCKEYWITHID, privateKey, OSSL_CRMF_PRIVATEKEYINFO),
+    ASN1_IMP_OPT(OSSL_CRMF_ENCKEYWITHID, identifier,
```
shouldn't be

mpeylo commented on crypto/crmf/crmf_asn.c:
```diff
@@ -0,0 +1,239 @@
+
+
+ASN1_SEQUENCE(OSSL_CRMF_OPTIONALVALIDITY) = {
+    ASN1_EXP_OPT(OSSL_CRMF_OPTIONALVALIDITY, notBefore, ASN1_TIME, 0),
```
should probably be ASN1_IMP_OPT due to the DEFINITIONS IMPLICIT TAGS ::=

mattcaswell replied:
Ok. This kinda makes sense.

mpeylo commented on crypto/crmf/crmf_asn.c:
```diff
@@ -0,0 +1,239 @@
+    ASN1_IMP_OPT(OSSL_CRMF_CERTTEMPLATE, signingAlg,   X509_ALGOR, 2),
+    ASN1_EXP_OPT(OSSL_CRMF_CERTTEMPLATE, issuer,       X509_NAME, 3),
+    ASN1_IMP_OPT(OSSL_CRMF_CERTTEMPLATE, validity,OSSL_CRMF_OPTIONALVALIDITY,4),
+    ASN1_EXP_OPT(OSSL_CRMF_CERTTEMPLATE, subject,      X509_NAME, 5),
```
Now that's the strange thing; that is because it doesn't work otherwise.

Here the Wireshark trace with EXP:

![image](https://user-images.githubusercontent.com/9685382/49833261-a5525b00-fda1-11e8-8690-cac8c3b68172.png)

Here with IMP:
![image](https://user-images.githubusercontent.com/9685382/49833323-cadf6480-fda1-11e8-9674-ad721c6fa8e6.png)

mattcaswell replied:
Name has no tagging in the PKIX1Explicit88 module:

````
Name ::= CHOICE { -- only one possibility for now --
      rdnSequence  RDNSequence }
````
So I guess it comes under the "use explicit for an untagged choice type" rule.

mpeylo commented on crypto/crmf/crmf_asn.c:
```diff
@@ -0,0 +1,239 @@
+IMPLEMENT_ASN1_DUP_FUNCTION(OSSL_CRMF_CERTREQUEST)
+
+
+ASN1_SEQUENCE(OSSL_CRMF_MSG) = {
```
I guess because the string is then used with suffixes in many function names, like "OSSL_CRMF_MSG_set_validity" - and that led to shorter function name strings. 

That could be changed if it is a requirement to always use the smae strangs as in ASN.1, while I think that OpenSSL usually is far less strict with that than the CMP/CRMF code.

similar for CertReqMessages / OSSL_CRMF_MSGS

mattcaswell replied:
Ok

mpeylo commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+    }
+
+    /* OpenSSL defaults all bit strings to be encoded as ASN.1 NamedBitList */
+    ps->signature->flags &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07);
```
I remember how I had to fight rather long to figure that out. That's somewhat burried deep in crypto/asn1/a_bitstr.c - the difference is that the OpenSSL ASN.1 encoder removes trailing 0s if that is not set as it handles BIT STRINGs per default as NamedBitString.

See ITU-T X.680:

> 22.7 When a "NamedBitList" is used in defining a bitstring type ASN.1 encoding rules are free to add (or remove) arbitrarily any trailing 0 bits to (or from) values that are being encoded or decoded. Application designers should therefore ensure that different semantics are not associated with such values which differ only in the number of trailing 0 bits.

Quite common to do that:

> $ grep -r '(ASN1_STRING_FLAG_BITS_LEFT | 0x07)'
> crypto/cmp/cmp_lib.c:    prot->flags &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07);
> crypto/x509/x_pubkey.c:        pub->public_key->flags &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07);
> crypto/dh/dh_ameth.c:        pubkey->flags &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07);
> crypto/asn1/a_sign.c:    signature->flags &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07);
> crypto/asn1/a_sign.c:    signature->flags &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07);
> crypto/asn1/asn1_gen.c:                &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07);
> crypto/asn1/a_bitstr.c:    ret->flags &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07); /* clear */
> crypto/asn1/a_bitstr.c:    a->flags &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07); /* clear, set on write */
> crypto/crmf/crmf_lib.c:    ps->signature->flags &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07);
> crypto/ec/ec_ameth.c:        pubkey->flags &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07);
> crypto/ec/ec_asn1.c:        curve->seed->flags &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07);
> crypto/ec/ec_asn1.c:        priv_key->publicKey->flags &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07);

DDvO replied:
I agree

DDvO commented on crypto/crmf/crmf_asn.c:
```diff
@@ -0,0 +1,239 @@
+ * Copyright Nokia 2007-2018
+ * Copyright Siemens AG 2015-2018
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
```
The next push will contain commit "replace 'OpenSSL license' by 'Apache License 2.0'"

DDvO commented on crypto/crmf/crmf_asn.c:
```diff
@@ -0,0 +1,239 @@
+
+/* explicit #includes not strictly needed since implied by the above: */
+#include <openssl/crmf.h>
+
```
Good point, reference will be included (also to RFC 4210 in cmp.h).

FdaSilvaYY commented on doc/man3/OSSL_CRMF_pbmp_new.pod:
```diff
@@ -0,0 +1,78 @@
+
+ OSSL_CRMF_PBMPARAMETER *pbm = NULL;
+ unsigned char *msg = "Hello";
+ unsigend char *sec = "SeCrEt";
```
`unsigend`

DDvO replied:
Fixed typos

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+
+int OSSL_CRMF_MSG_set0_SinglePubInfo(OSSL_CRMF_SINGLEPUBINFO *spi,
+                                     int method, GENERAL_NAME *nm) {
+    int error = CRMF_R_NULL_ARGUMENT;
```
Added such blank lines in all *.c files.

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+    if (spi == NULL)
+        goto err;
+    if (method < OSSL_CRMF_PUB_METHOD_DONTCARE ||
+        method > OSSL_CRMF_PUB_METHOD_LDAP) {
```
Thx; layout fixed.

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+                                 OSSL_CRMF_PKIPUBLICATIONINFO *pi,
+                                 OSSL_CRMF_SINGLEPUBINFO *spi) {
+    int error = ERR_R_MALLOC_FAILURE;
+    if (pi == NULL || spi == NULL) {
```
Fixed

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+        return 0;
+    }
+
+    if ((crm->regInfo) == NULL) {
```
Fixed

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+        return 0;
+    }
+
+    if ((tmpl->extensions) == NULL) {
```
Removed

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+    size_t crlen, max_sig_size;
+    unsigned int siglen;
+    unsigned char *crder = NULL, *sig = NULL;
+    int alg_nid=0, md_nid=0;
```
Fixed

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+ err:
+    CRMFerr(CRMF_F_CRMF_POPOSIGNINGKEY_INIT, CRMF_R_ERROR);
+    OPENSSL_free(crder);
+    if (ctx)
```
Removed

DDvO commented:
I've just finished my pass on all the recent review comments.
The only comments that I see still open are the two ones that @mpeylo wants to look into.

FdaSilvaYY commented on crypto/asn1/tasn_utl.c:
```diff
@@ -31,6 +31,11 @@ int asn1_get_choice_selector(ASN1_VALUE **pval, const ASN1_ITEM *it)
     int *sel = offset2ptr(*pval, it->utype);
     return *sel;
 }
+int asn1_get_choice_selector_const(const ASN1_VALUE **pval, const ASN1_ITEM *it)
```
Blank line between the two methods,

DDvO replied:
Done

FdaSilvaYY commented on crypto/cms/cms_io.c:
```diff
@@ -14,6 +14,7 @@
 #include <openssl/cms.h>
 #include "cms_lcl.h"
 
+/* unfortunately cannot constify BIO_new_NDEF() due to this funcion */
```
`funcion`

DDvO replied:
Fixed

FdaSilvaYY commented:
I fully support the API constification you made, but it deserves its own PR, IMHO, as it is large patch and not dependent of CRMF support added in this PR.

FdaSilvaYY commented on crypto/crmf/crmf_asn.c:
```diff
@@ -0,0 +1,239 @@
+#include <openssl/crmf.h>
+
+ASN1_SEQUENCE(OSSL_CRMF_PRIVATEKEYINFO) = {
+    ASN1_SIMPLE(OSSL_CRMF_PRIVATEKEYINFO, version, ASN1_INTEGER),
```
Why don't you use the ASN1_EMBED & co for these not-optional fields ?
It can spare various allocations and indirections.
Or you wish to keep compatibility with pre-1.1.0 code ?

mpeylo replied:
The relevant code was written far before ASN1_EMBED became available. It would be beneficial for us not to use it yet, as we're aiming to maintain compatibility with 1.0.2 for the time being.

mattcaswell commented:
There seems to be quite a number of my comments still outstanding. Also we need a rebase of this.

mattcaswell commented:
Are there CRMF specific tests?

DDvO commented:
> There seems to be quite a number of my comments still outstanding. 

Most of them were already done, while three comments regarding ASN.1, two by you/@mattcaswell  and one by @FdaSilvaYY, are open (to be answered by @mpeylo).

DDvO commented:
> Also we need a rebase of this.

Indeed. Hope @mpeylo can do this soon.

mspncp commented:
> Thanks @mspncp for this hit!
> I'll follow it as soon as this PR has been rebased on the current master.

I noticed that you have two constifying commits, 168824237f and b1a88522c3. So it's probably easier to reorder the commits of `cmp_incremental2` first, doing an interactive rebase such that these two commits are moved to the front of the rebase-todo-list and squashed into a single commit. After the rebase, you can simply create a `constify` topic branch from the first commit of `cmp_incremental2`.


```
git checkout cmp_incremental2
git rebase -i master
```

The rebase-todo list appears. Place the constify commit and its fixup at the beginning as follows:

```
pick 168824237f constify *_dup() and *i2d_*() and related functions as far as possible
fixup b1a88522c3 constify *_dup() and *i2d_*() and related functions as far as possible: two nits
pick cf6374b823 Certificate Management Protocol (CMP, RFC 4210) extension to OpenSSL Also includes CRMF (RFC 4211) and HTTP transfer (RFC 6712)
pick e4c985090a a couple of minor doc improvements according to API review by @bernd-edlinger
pick bad83c1d8d remove OSSL_CRMF_ASN1_get_int() due to review comments
 ...

```

The rebasing will stop with a conflict in `util/libcrypto.num`. Discard the incoming changes and rerun `make update`:

```
git checkout --ours util/libcrypto.num
git add util/libcrypto.num
make update
```

Now check the diffs in `util/libcrypto.num` and after that, continue rebasing.
```
git diff
git add -u
git rebase --continue
```

DDvO commented:
> I noticed that you have two constifying commits, [1688242](https://github.com/openssl/openssl/commit/168824237fd33e634b9e3f021b8404ea758e5871) and [b1a8852](https://github.com/openssl/openssl/commit/b1a88522c3c48690f21d75c62931fba803d34bea). So it's probably easier to reorder the commits of `cmp_incremental2` first, doing an interactive rebase such that these two commits are moved to the front of the rebase-todo-list and squashed into a single commit.

Excellent hint - thanks! I've been needing such a technique for quite a while.
Done this for the two commits:
```
git rebase -i cb659e3fac3717ba8f92e8a7c1795655028b81cf
```
and (for now) force-pushed the result here.

mspncp commented:
When the merge conflict is resolved, you can continue the rebase operation:

```
git rebase --continue
```

Sometimes git will complain, that the patch is empty. (This is the case if you reverted all changes.). Then just skip the patch:

```
git rebase --skip
```

This procedure will have to be repeated several times.

Alternatively, if you agree that the pr can be squashed into a single commit at the end, you don't have to run `Configure ... ;  make update` at every conflicting commit. Just revert libcrypto.num every time git stops at a merge conflict

```
git checkout --ours util/libcrypto.num
git add util/libcrypto.num
git rebase --continue
```

And run `Configure ... ; make update` once at the very end of the rebase.

mspncp commented:
Finally, you need to force-push your changes, because you local branch has moved.

mspncp commented:
Sorry, the last two posts were intended for #7893. I deleted them again.

DDvO commented:
Meanwhile @mpeylo and me have reacted on all remaining open comments and pushed respective changes.

Does this satisfy all changes requested? 
Are there maybe further comments?

DDvO commented:
Rebased this PR, which became necessary because is its based on https://github.com/openssl/openssl/pull/8029, which needed to be rebased after recent rearrangements of ESS code.

HBrock commented:
> Does this satisfy all changes requested?
> Are there maybe further comments?

@mattcaswell are there further changes needed for chunk 2?

DDvO commented:
I had to patch `util/libcrypto.num` manually because my local runs of `make update`
kept producing different numbers for the CRMF ASN.1 `_it` functions than Travis CI, 
where this produces, e.g.,
```
OSSL_CRMF_ENCRYPTEDVALUE_it 4594 3_0_0 EXIST:!EXPORT_VAR_AS_FUNCTION:VARIABLE:CMP,CRMF
OSSL_CRMF_ENCRYPTEDVALUE_it 4594 3_0_0 EXIST:EXPORT_VAR_AS_FUNCTION:FUNCTION:CMP,CRMF
```
I failed to find out where that strange discrepancy comes from.

DDvO commented:
As mentioned, all reviewer comments have been dealt with.

Any more changes to do on this PR?

mattcaswell commented:
So, is my understanding correct that this builds on top of #8029? If so it would probably be wise to get that one finished and merged to simplify the review of this one.

DDvO commented:
Yes, it's based on that PR #8029, but reviews of this PR can be done pretty independently because they can focus on the new files in `crypto\crmf`. 
The first commit (which is PR #8029) is present just to make sure the CRMF code compiles as given here,
and its influence is just that it makes one or two additional `const` parameter types possible (without having to cast to `const`) in this PR.

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+
+    return 1;
+ oom:
+    CRMFerr(CRMF_F_OSSL_CRMF_MSG_PUSH0_REGCTRL, ERR_R_MALLOC_FAILURE);
```
I'd say yes, because `OPENSSL_sk_push()` fails with the provided stack is NULL (which cannot be the case here) or memory is insufficient to grow the stack (as far as needed).

BTW, it's a pity that one still cannot rely on `OPENSSL_malloc()` and friends throwing  `ERR_R_MALLOC_FAILURE` in case of lack of memory, which would relieve all callers from (trying to find out what went wrong and) doing that. See https://github.com/openssl/openssl/issues/6251

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+    }
+
+    if ((tmpl = OSSL_CRMF_MSG_get0_tmpl(crm)) == NULL)
+        goto oom;
```
Ah, right. Fixed this here and in other places.

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+        return 0;
+    }
+    if ((tmpl = OSSL_CRMF_MSG_get0_tmpl(crm)) == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_MSG_SET0_EXTENSIONS, ERR_R_MALLOC_FAILURE);
```
Yes, fixed also here.

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+int OSSL_CRMF_MSG_set0_extensions(OSSL_CRMF_MSG *crm,
+                                  X509_EXTENSIONS *exts)
+{
+    OSSL_CRMF_CERTTEMPLATE *tmpl;
```
Fixed

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+}
+
+/* get ASN.1 encoded integer, return -1 on error */
+static int CRMF_ASN1_get_int(int func, const ASN1_INTEGER *a)
```
For the given use cases, 32-bit integers are more than sufficient.
I think that the two range checks done here do not hurt. and
I'd consider it not worth the effort consistently changing this and all related uses to `int64_t`.

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+        goto err;
+    crlen = (size_t)len;
+
+    max_sig_size = EVP_PKEY_size((EVP_PKEY *)pkey);
```
Thanks

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+        exts = NULL; /* do not include empty extensions list */
+    }
+
+    tmpl->extensions = exts;
```
Indeed we should do that since the idea is that `OSSL_CRMF_MSG_set0_extensions()` consumes the given extensions. Done.

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+        goto err;
+    }
+    if (!(OBJ_find_sigid_algs(alg_nid, &md_nid, NULL)
+          && (alg = EVP_get_digestbynid(md_nid)) != NULL)) {
```
Changed

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+        goto err;
+    if (!(EVP_SignUpdate(ctx, crder, crlen)))
+        goto err;
+    if (!(EVP_SignFinal(ctx, sig, &siglen, (EVP_PKEY *)pkey)))
```
I've replaced the calls of EVP_Sign*() by EVP_DigestSign*()
and removed the unsafe cast to `const` (which is not really needed here).

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+ err:
+    CRMFerr(CRMF_F_CRMF_POPOSIGNINGKEY_INIT, CRMF_R_ERROR);
+    OPENSSL_free(crder);
+    if (ctx)
```
Done

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,767 @@
+    OPENSSL_free(sig);
+    return 1;
+ err:
+    CRMFerr(CRMF_F_CRMF_POPOSIGNINGKEY_INIT, CRMF_R_ERROR);
```
Good points - done. 
(I never would have introduced such overlaps and redundancies.)

DDvO commented:
Gosh, I just realized that a large part of the comments of Dec 11 really were still outstanding!
I overlooked them by now because the GitHub page had folded their view.
Thanks @mattcaswell for all your thoughtful review comments! Finally I've dealt with the remaining ones.

DDvO commented:
BTW, so far I had been reluctant to use the "Resolve Conversation" button after handling a comment, since I thought I cannot be sure that my solution is accepted. 
But meanwhile I have the impression that I should have used it for all such cases to signal to the respective reviewer that I believe that the issue is solved, right?

mattcaswell commented:
> But meanwhile I have the impression that I should have used it for all such cases to signal to the respective reviewer that I believe that the issue is solved, right?

I think its a useful way of indicating that we think an issue is solved so we can easily see what is still outstanding. As a reviewer I always check that I'm happy with the response anyway and reopen it if I'm not.

DDvO commented:
I've rebased this PR to solve minor conflicts with two auto-generated files.

@mattcaswell, does this PR meanwhile look good to you, or is there more to do?

mattcaswell commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,755 @@
+        }
+        return 1;
+    case OSSL_CRMF_POPO_KEYENC:
+#ifdef OSSL_CMP_CERTREP_NEW_IMPLEMENTS_POPO_KEYENC /* TODO */
```
What is this?

DDvO replied:
This code is just a reminder what to do if the 'encrypted cert' method is implemented in `OSSL_CMP_certrep_new()` (which is not part of this PR).
I've re-phrased it as a simpler TODO comment.

mattcaswell commented on crypto/crmf/crmf_pbm.c:
```diff
@@ -0,0 +1,220 @@
+        CRMFerr(CRMF_F_OSSL_CRMF_PBM_NEW, CRMF_R_UNSUPPORTED_ALGORITHM);
+        goto err;
+    }
+    if (HMAC(m, basekey, bklen, msg, msglen, mac_res, maclen) != NULL)
```
Can we use EVP_MAC here?

mpeylo replied:
Yes we can - although it's not obvious what the instant benefit might be.

On it's way, pending review https://github.com/mpeylo/cmpossl/pull/163

DDvO replied:
Done, taking over @mpeylo's code.

mattcaswell replied:
> Yes we can - although it's not obvious what the instant benefit might be.

That API will be deprecated in OpenSSL 3. The EVP APIs will benefit from the new Provider architecture (https://www.openssl.org/docs/OpenSSL300Design.html)...so for example, a user has the ability to mandate that all crypto goes through the new FIPS module.

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,755 @@
+        goto err;                                                         \
+    if ((atav->type = OBJ_nid2obj(NID_id_##ctrlinf##_##atyp)) == NULL)    \
+        goto err;                                                         \
+    if ((atav->value.atyp = valt##_dup((valt *)in)) == NULL)              \
```
Good point, not needed anymore :)

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,755 @@
+    if ((cid = OSSL_CRMF_CERTID_new()) == NULL)
+        goto oom;
+
+    if (!X509_NAME_set(&cid->issuer->d.directoryName, (X509_NAME *)issuer))
```
Also here, the cast is not needed anymore; removed.

DDvO commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,755 @@
+        new = 1;
+    }
+
+    if (!sk_X509_EXTENSION_push(tmpl->extensions, (X509_EXTENSION *)ext))
```
Yes, the cast is valid, which should be intuitively clear.
Unfortunately it is still needed because the following decl in `safestack.h`:
```
    static ossl_unused ossl_inline int sk_##t1##_push(STACK_OF(t1) *sk, t2 *ptr) \
    { \
        return OPENSSL_sk_push((OPENSSL_STACK *)sk, (const void *)ptr); \
    } \
```
still needlessly does not declare `t2 *ptr` as `const`.
I'm reluctant to correct it in https://github.com/openssl/openssl/pull/8029 because that PR is already too large and this would require to renew its first approval.

mattcaswell replied:
I agree.

DDvO commented:
I've just rebased this PR and removed all remnants of `ifndef NO_CRMF` (which are superseded by `NO_CMP`).
@mattcaswell, are you fine with the current state, where I've handled all your (recent) comments?

DDvO commented:
Thanks @mattcaswell

mattcaswell commented:
Now that #8029 is in, please can you rebase this?

DDvO commented:
> Now that #8029 is in, please can you rebase this?

With pleasure, done!

mattcaswell commented:
Thanks! So we still need a second reviewer for this. This PR still depends on #7328, which has received approval but can't go in until this one is also approved. #7328 covered the documentation and external header file changes so those don't need to be reviewed again.

bernd-edlinger commented:
Okay, I'll have a look-

bernd-edlinger commented on crypto/crmf/crmf_int.h:
```diff
@@ -0,0 +1,387 @@
+ *  -- used if the RA has already verified that the requester is in
+ *  -- possession of the private key
+ *  signature                 [1] POPOSigningKey,
+ *  keyEncipherment   [2] POPOPrivKey,
```
sloppy formatting.

DDvO replied:
Fixed

bernd-edlinger commented on crypto/crmf/crmf_int.h:
```diff
@@ -0,0 +1,387 @@
+ * publicKey        [6] SubjectPublicKeyInfo      OPTIONAL,
+ * issuerUID        [7] UniqueIdentifier          OPTIONAL,
+ * subjectUID   [8] UniqueIdentifier              OPTIONAL,
+ * extensions   [9] Extensions                    OPTIONAL }
```
formatting.

DDvO replied:
Fixed

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+ * CRMF implementation by Martin Peylo, Miikka Viljanen, and David von Oheimb.
+ */
+
+/* NAMING
```
multi line comments start with /*

DDvO replied:
See above

bernd-edlinger replied:
I meam why not:
```
 /*
  * NAMING
  * 
  *  The 0 functions ....
```

DDvO replied:
changed

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+    }
+
+    if (crm->certReq->controls == NULL) {
+        if ((crm->certReq->controls =
```
do we really need to do an assignment in an if statement?

DDvO replied:
Well, it's not really needed this way, but it saves repeating the (in this case a bit lengthy) lvalue expression.
Changed anyway.

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+                                     int method, GENERAL_NAME *nm)
+{
+    if (spi == NULL
+        || method < OSSL_CRMF_PUB_METHOD_DONTCARE
```
add 4 spaces alignment here

DDvO replied:
Done

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+
+    if (!ASN1_INTEGER_set(spi->pubMethod, method))
+        return 0;
+    spi->pubLocation = nm;
```
what if pubLocation is != NULL shouldn't it be freed first?

DDvO replied:
Good point - as also implied by its documentation, the function is responsible for freeing that name structure, so it should free any pre-existing value. Done.

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+        return 0;
+    }
+    if (pi->pubinfos == NULL)
+        if ((pi->pubinfos = sk_OSSL_CRMF_SINGLEPUBINFO_new_null()) == NULL)
```
I would prefer that if in an extra statement.

DDvO replied:
Done

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+{
+    if (pi == NULL
+        || action < OSSL_CRMF_PUB_ACTION_DONTPUBLISH
+        || action > OSSL_CRMF_PUB_ACTION_PLEASEPUBLISH) {
```
indentations

DDvO replied:
changed

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+    }
+
+    if (crm->regInfo == NULL) {
+        if ((crm->regInfo =
```
we have other places where we do it this way:
```C
OSSL_CRM_MSG *info = NULL;

if (crm->regInfo == NULL)
    crm->regInfo = info = xxx_new();

if (crm->regInfo == NULL)
    goto oom;

...
oom:
   if (info != NULL)
       crm->regInfo = NULL;
   xxx_free(info);
```

DDvO replied:
Nice pattern; changed like this.

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+int OSSL_CRMF_MSG_get_certReqId(OSSL_CRMF_MSG *crm)
+{
+    if (crm == NULL
+        || /* not really needed: */ crm->certReq == NULL) {
```
indentation

DDvO replied:
changed

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+}
+
+/* get ASN.1 encoded integer, return -1 on error */
+static int CRMF_ASN1_get_int(int func, const ASN1_INTEGER *a)
```
for a local function the name should be lowercase, like crmf_asn1_get_int

DDvO replied:
changed

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+        goto err;
+    }
+    if (!OBJ_find_sigid_algs(alg_nid, &md_nid, NULL)
+        || (alg = EVP_get_digestbynid(md_nid)) == NULL) {
```
indentation

DDvO replied:
changed

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+        || (ctx = EVP_MD_CTX_new()) == NULL
+        || EVP_DigestSignInit(ctx, NULL, alg, NULL, pkey) <= 0
+        || EVP_DigestSignUpdate(ctx, crder, crlen) <= 0
+        || EVP_DigestSignFinal(ctx, NULL, &siglen) <= 0) {
```
indentation

DDvO replied:
changed

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+        goto err;
+    }
+    if (EVP_DigestSignFinal(ctx, sig, &siglen) <= 0
+        || !ASN1_BIT_STRING_set(ps->signature, sig, siglen)) {
```
indentation

DDvO replied:
changed

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+        if ((pp->value.keyEncipherment->value.subsequentMessage =
+             ASN1_INTEGER_new()) == NULL
+            ||
+            !ASN1_INTEGER_set(pp->value.keyEncipherment->value.subsequentMessage,
```
there must be a way to simplify this, maybe declar a pointer to 
pp->value.keyEncipherment->value.subsequentMessage
and use that ?

DDvO replied:
Indeed - done.

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+    if (a == b)
+        return 0;
+    if (a == NULL
+        || !X509_PUBKEY_get0_param(NULL, NULL, NULL, &algA, a) || algA == NULL)
```
indentation

DDvO replied:
changed

bernd-edlinger commented on crypto/crmf/crmf_pbm.c:
```diff
@@ -0,0 +1,230 @@
+    OPENSSL_free(mac_res);
+
+    if (pbmp != NULL && pbmp->mac != NULL) {
+        char buf[128];
```
empty line after declaraions

DDvO replied:
done

bernd-edlinger commented on crypto/err/openssl.txt:
```diff
@@ -348,6 +348,28 @@ CONF_F_NCONF_NEW:111:NCONF_new
 CONF_F_SSL_MODULE_INIT:123:ssl_module_init
 CONF_F_STR_COPY:101:str_copy
+CRMF_F_CRMF_POPOSIGNINGKEY_INIT:100:CRMF_poposigningkey_init
+CRMF_F_OSSL_CRMF_CERTID_GEN:101:OSSL_CRMF_CERTID_gen
```
the numbers are already in use, is that a problem?

mpeylo replied:
openssl.txt needs to be auto-generated based on latest version in master after every re-base.

DDvO replied:
The values are auto-generated and start with `100` for each sub-library.
It is fine that the values overlap with other sub-libs since they are distinguished by the module identifiers like `ERR_LIB_CRMF` and `ERR_LIB_CONF`.

DDvO commented on crypto/crmf/crmf_asn.c:
```diff
@@ -0,0 +1,239 @@
+    ASN1_EXP(OSSL_CRMF_POPOSIGNINGKEYINPUT_AUTHINFO, value.sender,
+             GENERAL_NAME, 0),
+    ASN1_SIMPLE(OSSL_CRMF_POPOSIGNINGKEYINPUT_AUTHINFO, value.publicKeyMAC,
+            OSSL_CRMF_PKMACVALUE)
```
Yep, fixed

DDvO commented on crypto/crmf/crmf_asn.c:
```diff
@@ -0,0 +1,239 @@
+
+ASN1_SEQUENCE(OSSL_CRMF_POPOSIGNINGKEYINPUT) = {
+    ASN1_SIMPLE(OSSL_CRMF_POPOSIGNINGKEYINPUT, authinfo,
+            OSSL_CRMF_POPOSIGNINGKEYINPUT_AUTHINFO),
```
fixed

DDvO commented on crypto/crmf/crmf_asn.c:
```diff
@@ -0,0 +1,239 @@
+
+ASN1_SEQUENCE(OSSL_CRMF_POPOSIGNINGKEY) = {
+    ASN1_IMP_OPT(OSSL_CRMF_POPOSIGNINGKEY, poposkInput,
+                 OSSL_CRMF_POPOSIGNINGKEYINPUT,0),
```
Yep, added

DDvO commented on crypto/crmf/crmf_asn.c:
```diff
@@ -0,0 +1,239 @@
+     */
+    ASN1_IMP_OPT(OSSL_CRMF_CERTTEMPLATE, signingAlg,   X509_ALGOR, 2),
+    ASN1_EXP_OPT(OSSL_CRMF_CERTTEMPLATE, issuer,       X509_NAME, 3),
+    ASN1_IMP_OPT(OSSL_CRMF_CERTTEMPLATE, validity,OSSL_CRMF_OPTIONALVALIDITY,4),
```
The line had been squeezed into 80 chars; I've split the line now, adding space after ','.

DDvO commented on crypto/crmf/crmf_err.c:
```diff
@@ -0,0 +1,104 @@
+
+static const ERR_STRING_DATA CRMF_str_reasons[] = {
+    {ERR_PACK(ERR_LIB_CRMF, 0, CRMF_R_BAD_PBM_ITERATIONCOUNT),
+    "bad pbm iterationcount"},
```
I don't understand what you mean here, but this stuff is auto-generated by `mkerr.pl`. 
It should be correct by construction, and it works :)

bernd-edlinger replied:
OK

DDvO commented on crypto/crmf/crmf_err.c:
```diff
@@ -0,0 +1,104 @@
+int ERR_load_CRMF_strings(void)
+{
+#ifndef OPENSSL_NO_ERR
+    if (ERR_func_error_string(CRMF_str_functs[0].error) == NULL) {
```
It's always like this, as auto-generated by `mkerr.pl`.

DDvO commented on crypto/crmf/crmf_int.h:
```diff
@@ -0,0 +1,387 @@
+ */
+typedef struct OSSL_crmf_privatekeyinfo_st {
+    ASN1_INTEGER *version;
+    X509_ALGOR *AlgorithmIdentifier;
```
Good point, `privateKeyAlgorithm` is the field name, while `AlgorithmIdentifier` is the field type,
as given in https://tools.ietf.org/html/rfc4211#section-4.2.1
```
      PrivateKeyInfo ::= SEQUENCE {
         version                   INTEGER,
         privateKeyAlgorithm       AlgorithmIdentifier,
         privateKey                OCTET STRING,
         attributes                [0] IMPLICIT Attributes OPTIONAL
      }
```
Corrected.

DDvO commented on crypto/crmf/crmf_int.h:
```diff
@@ -0,0 +1,387 @@
+    ASN1_BIT_STRING *encValue;
+} /* OSSL_CRMF_ENCRYPTEDVALUE */;
+
+/*-
```
Not according to https://www.openssl.org/policies/codingstyle.html
which says
```
The preferred style for long (multi-line) comments is:

    /*-
     * This is the preferred style for multi-line
     * comments in the OpenSSL source code.
     * Please use it consistently.
     *
     * Description:  A column of asterisks on the left side,
     * with beginning and ending almost-blank lines.
     */
```

mattcaswell replied:
The coding style goes on to say, immediately after that example:
`Note the initial hyphen to prevent indent from modifying the comment.
Use this if the comment has particular formatting that must be preserved.`

So the hyphen is optional, and should be used if there is a "formatted" comment. Doesn't seem particularly the case here. But I also don't think it is "against" the coding style either.

bernd-edlinger replied:
It is just 90% of all multiline comments don't do that, but if you do it consistently I won't object.

DDvO replied:
Thanks for pointing to those two lines in the guidelines which I had overlooked.

Many of the multi-line comments in crmf_int.h do have special formatting inherited from data structure definitions in RFC 4211. For many of those the hyphen was actually missing, so I've just added it there, as well as in some cases in crmf_*.c where special formatting (when documenting static functions) was intended. On the other hand, many of the comments in the CRMF source files do not have special formatting, so I've removed the hyphen there.

DDvO commented on crypto/crmf/crmf_int.h:
```diff
@@ -0,0 +1,387 @@
+} OSSL_CRMF_PRIVATEKEYINFO;
+DECLARE_ASN1_FUNCTIONS(OSSL_CRMF_PRIVATEKEYINFO)
+
+/*-
```
See above

DDvO commented on crypto/crmf/crmf_int.h:
```diff
@@ -0,0 +1,387 @@
+} OSSL_CRMF_ENCKEYWITHID;
+DECLARE_ASN1_FUNCTIONS(OSSL_CRMF_ENCKEYWITHID)
+
+/*-
```
See above

DDvO commented on crypto/crmf/crmf_int.h:
```diff
@@ -0,0 +1,387 @@
+ */
+struct OSSL_crmf_pkipublicationinfo_st {
+    ASN1_INTEGER *action;
+    OSSL_CRMF_PUBINFOS *pubinfos;
```
Yes - corrected.

DDvO commented on crypto/crmf/crmf_int.h:
```diff
@@ -0,0 +1,387 @@
+DECLARE_ASN1_FUNCTIONS(OSSL_CRMF_POPOSIGNINGKEYINPUT_AUTHINFO)
+
+typedef struct OSSL_crmf_poposigningkeyinput_st {
+    OSSL_CRMF_POPOSIGNINGKEYINPUT_AUTHINFO *authinfo;
```
Good spotting - corrected.

DDvO commented on crypto/crmf/crmf_int.h:
```diff
@@ -0,0 +1,387 @@
+ * POPOSigningKey ::= SEQUENCE {
+ *  poposkInput               [0] POPOSigningKeyInput OPTIONAL,
+ *  algorithmIdentifier   AlgorithmIdentifier,
+ *  signature                         BIT STRING }
```
Done

DDvO commented:
> mostly indentation nits

Thanks @bernd-edlinger for your very swift reviewing pass. 
I've just answered all comments and done all due changes.

bernd-edlinger commented on crypto/crmf/crmf_int.h:
```diff
@@ -0,0 +1,394 @@
+} /* OSSL_CRMF_CERTID */;
+DECLARE_ASN1_DUP_FUNCTION(OSSL_CRMF_CERTID)
+
+/*
```
I mean here it is again other way round without any reason.

DDvO replied:
Indeed, changed

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,751 @@
+ */
+
+/*
+ * This file contains the functions that handle the individual items inside
```
would it make sinse to move this comment before the prevoius?

DDvO replied:
Yes, I just had the same thought; done.

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,751 @@
+                CRMF_R_NULL_ARGUMENT);
+        return 0;
+    }
+    if (pi->pubInfos == NULL)
```
this needs braces.

DDvO replied:
Actually in this special case the braces are not critical, while the indentation was wrong/misleading.

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,751 @@
+        return 0;
+    }
+
+    if (sk_X509_EXTENSION_num(exts) == 0) {
```
what if ests is NULL?

DDvO replied:
If `exts == NULL` then `sk_X509_EXTENSION_num(exts) == -1` and nothing needs to be free'd.

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,751 @@
+
+    /* first the symmetric key needs to be decrypted */
+    if ((pkctx = EVP_PKEY_CTX_new(pkey, NULL)) != NULL
+        && EVP_PKEY_decrypt_init(pkctx)) {
```
indentation,
but OTOH how about?
```C
  pkctx = EVP_PKEY_CTX_new(pkey, NULL);
  if (pkctx != NULL && && EVP_PKEY_decrypt_init(pkctx)) {
```

I think this if (x=y) makes sometimes sense but can also be a habit.

DDvO replied:
I agree, and I've changed it accordingly (just not using two times `&&`).

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+/*
+ * NAMING
+ *
+ * The 0 versions use the supplied structure pointer directly in the parent and
```
I wonder if you want to use the term "The 0 functions use.."
because below you say "The 1 functions used..." ?

DDvO replied:
Yes, "0 functions" is better; changed.

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+}
+
+
+/*-
```
this is a normal comment, thus no /*- ?

DDvO replied:
I'd say this is a case of special formatting since each comment line is semantically one statement/sentence.
If the comment was re-formatted (with newlines elsewhere), it would become hard to parse/understand.

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+static IMPLEMENT_ASN1_DUP_FUNCTION(X509_PUBKEY)
+IMPLEMENT_CRMF_CTRL_FUNC(protocolEncrKey, X509_PUBKEY, regCtrl)
+
+/*-
```
no formatted text?

DDvO replied:
Same as above.
Anyway, as discussed the distinction is marginal, so if you recommend changing it I can do of course.

bernd-edlinger replied:
Well, okay, I just want to make sure that everything is consistent.
But if you are comfortable this way, it's fine for me.

bernd-edlinger commented on crypto/crmf/crmf_pbm.c:
```diff
@@ -0,0 +1,231 @@
+    }
+
+    /* compute the basekey of the salted secret */
+    if (!(EVP_DigestInit_ex(ctx, m, NULL)))
```
the inner parentheses are unnecessary, in several places below.

DDvO replied:
removed

DDvO commented:
I've just rebased this PR due to ever-annoying conflicts in the auto-generated `util/libcrypto.num`.

DDvO commented:
@bernd-edlinger, do you see anything more to do after the recent minor improvements, or can you meanwhile approve this?

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+        if (tag == NULL
+                || !ASN1_INTEGER_set(tag, OSSL_CRMF_SUBSEQUENTMESSAGE_ENCRCERT))
+            goto oom;
+        pp->value.keyEncipherment->value.subsequentMessage = tag;
```
resource leak.
please move this assignment before the if. If you like combine it with pp->...subsequentMessage = tag = ...new();

bernd-edlinger replied:
.. or explicitly free tag like the ps above.

DDvO replied:
Good spotting! I've moved the assignment (even one more further up, to keep the two assignments to `pp->value.keyEncipherment->type` adjacent).
Combining the two assignments would have been nice but would have gone beyond the 80-char line limit.

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+    if (pi->pubInfos == NULL)
+        goto oom;
+
+    if (!(sk_OSSL_CRMF_SINGLEPUBINFO_push(pi->pubInfos, spi)))
```
inner parentheses are unnecessary.

DDvO replied:
removed

bernd-edlinger commented:
I wanted to approve already, but found a few things I overlooked before.

DDvO commented:
> I wanted to approve already, but found a few things I overlooked before.

I've meanwhile handled all of them, and Travis CI and AppVeyor have passed again :)

DDvO commented:
Very pleased that this is now fully approved :-)
Thanks, also in the name of @mpeylo, 
to all reviewers: @bernd-edlinger, @FdaSilvaYY, @mattcaswell, and @mspncp.

mattcaswell commented:
Woop! Pushed! Thanks all!

So...chunk 3 now?

DDvO commented:
> So...chunk 3 now?

Basically yes :)

Martin and me just need some time now to rebase our CMP code on this 
and consolidate some internal feature branches before we can push the next chunk.

And I think we need to clarify first what exactly is going to be part of chunk 3:
it could be the full CMP API, i.e., all CMP header files, yet due to their size I suppose it will be helpful to break this up, starting for instance with a lower-level portion regarding basic CMP messages handling. Even this is going to be larger than the CRMF implementation.

mattcaswell commented:
Choose the smallest self-contained piece that you can find. Ideally one that can be committed as-is so that we don't have multiple PRs in-flight that depend on each other.

DDvO commented:
> So...chunk 3 now?

Here it is:  #8669

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+        if (EVP_PKEY_decrypt(pkctx, NULL, &eksize, encKey->data, encKey->length)
+                <= 0
+                || (ek = OPENSSL_malloc(eksize)) == NULL
+                || EVP_PKEY_decrypt(pkctx, ek, &eksize, encKey->data,
```
you don't use eksize, you dont cleat the ek which is secret, right?

bernd-edlinger replied:
is the encKey->data authenticated, or could this be used for a padding oracle?

Akretsch replied:
Fixed: eksize is no used for clearing ek in #9774 . 
If CRMF (and therefore encSymmKey  as part of EncryptedValue) is used in context of a CMP PKIMessage, the PKIMessage has to be protected in the most use cases. Protections means applying password based or signature based message integrity. See RFC4210, chapter 5.1.3 .

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+        goto oom;
+    EVP_CIPHER_CTX_set_padding(evp_ctx, 0);
+
+    if (!EVP_DecryptInit(evp_ctx, cipher, ek, iv)
```
you don't check that eksize is equal to EVP_CIPHER_key_length
or try to set EVP_CIPHER_CTX_set_key_length in case it is a variable key length?

Akretsch replied:
Added the check in #9774

bernd-edlinger commented on crypto/crmf/crmf_lib.c:
```diff
@@ -0,0 +1,753 @@
+    EVP_PKEY_CTX_free(pkctx);
+    OPENSSL_free(outbuf);
+    EVP_CIPHER_CTX_free(evp_ctx);
+    OPENSSL_free(ek);
```
OPENSSL_clear_free?

Akretsch replied:
Fixed in #9774
